src/SFA.DAS.ApprenticeAccounts/Application/Commands/UpdateAllApprenticePreferencesCommand/UpdateAllApprenticePreferencesCommandHandler.cs

```diff
@@ -0,0 +1,68 @@
+        private readonly ILogger<UpdateAllApprenticePreferencesCommandHandler> _logger;
+
+        public UpdateAllApprenticePreferencesCommandHandler(IApprenticePreferencesContext apprenticePreferencesContext,
+            IApprenticeContext apprenticeContext, IPreferencesContext preferencesContext, ILogger<UpdateAllApprenticePreferencesCommandHandler> logger)
+        {
+            _apprenticePreferencesContext = apprenticePreferencesContext;
+            _apprenticeContext = apprenticeContext;
+            _preferencesContext = preferencesContext;
+            _logger = logger;
+        }
+
+        public async Task<Unit> Handle(UpdateAllApprenticePreferencesCommand request, CancellationToken cancellationToken)
+        {
+            foreach (var apprenticePreference in request.ApprenticePreferences)
+            {
+                _logger.LogInformation($"Fetch Apprentice by Id. Id used: {apprenticePreference.ApprenticeId}");
+                var apprentice = await _apprenticeContext.Entities.FindAsync(apprenticePreference.ApprenticeId);
+
+                _logger.LogInformation($"Fetch Preference by Id. Id used: {apprenticePreference.PreferenceId}");
+                var preference = await _preferencesContext.Entities.FindAsync(apprenticePreference.PreferenceId);
+
+                if (apprentice == null || preference == null)
+                {
+                    _logger.LogError($"No Apprentice record found, or no Preference record found, or neither record found. Apprentice Id used; {apprenticePreference.ApprenticeId}, Preference Id used: {apprenticePreference.PreferenceId}");
+                    throw new InvalidOperationException();
+                }
+
+                _logger.LogInformation($"Fetch ApprenticePreferences record by Apprentice Id and Preference Id. Apprentice Id used: {apprenticePreference.ApprenticeId}, Preference Id used: {apprenticePreference.PreferenceId}");
+                var record =
+                    await _apprenticePreferencesContext.GetApprenticePreferenceForApprenticeAndPreference(
+                        apprenticePreference.ApprenticeId, apprenticePreference.PreferenceId);
+
+                if (record == null)
+                {
+                    await _apprenticePreferencesContext.AddAsync(new ApprenticePreferences(
+                        apprenticePreference.ApprenticeId, apprenticePreference.PreferenceId,
+                        apprenticePreference.Status, DateTime.Now, DateTime.Now));
+                    _logger.LogDebug($"Successfully created a new ApprenticePreferences record with Apprentice Id: {apprenticePreference.ApprenticeId} and Preference Id: {apprenticePreference.PreferenceId}");
+                }
+                else
+                {
+                    record.Status = apprenticePreference.Status;
+                    record.UpdatedOn = DateTime.Now;
+                    
+                    _apprenticePreferencesContext.Update(record);
```

Comment:
Same on the update if we can switch to save changes for ease ( and specifically if we can do it as a batch save changes on all the entities tracked rather than multiple individual updates. )